### PR TITLE
doc(xcy): polish mcts bot comments

### PR DESCRIPTION
**Pull request by HarryXuancy**
- polish mcts bot comments

### CODE REVIEW

File: zoo/board_games/mcts_bot.py [hunk 1]
```diff
@@ -5,23 +5,39 @@
...
+                        where 0 stands for position that has not been played yet, 
+                        1 represents a position that has been played by player 1.
+                        2 represents a position taken by player 2.
+                    - env.players = [1,2], representing two players, player 1 and player 2.
+                    - env.start_player_index = 0 or 1. It will be reset every round.
+                    - env._current_player = env.players[env.start_player_index], representing which player 
+                        should make a move in this round.
+                - The methods in this object implement functionalities such as environment transitions and obtaining game results.
+            #!!!!!!!为什么回合开始不直接重置current player!!!!!!!!!
+            - parent(): The parent node of current node. It is used for backpropagation. For the root node, this parameter is None.
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:35]: - parent (:obj:`MCTSNode`):  This parent node of the current node. The parent node is primarily used for backpropagation during the Monte Carlo Tree Search. For the root node, this parent returns None as it does not have a parent node.

File: zoo/board_games/mcts_bot.py [hunk 2]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
         pass
 
     def is_fully_expanded(self):
+        """
+        Overview:
+            Check whether the node is fully expanded. Whenever a new child node is selected first time, an element 
+            is popped out from the list of legal actions. When the list becomes empty, it indicates that all child 
+            nodes have been selected. This means that the parent node has been completed expanded.
+        """
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:85]:
```
This method checks if the node is fully expanded. A node is considered fully expanded when all of its child nodes have been selected at least once. 

Whenever a new child node is selected for the first time, a corresponding action is removed from the list of legal actions. Once the list of legal actions is depleted, it signifies that all child nodes have been selected, thereby indicating that the parent node is fully expanded.
```

File: zoo/board_games/mcts_bot.py [hunk 3]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
...
         return len(self.legal_actions) == 0
 
     def best_child(self, c_param=1.4):
         '''
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:89]: 使用"""

File: zoo/board_games/mcts_bot.py [hunk 4]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
...
+                - c is a parameter which balance exploration and exploitation.
         '''
+        # Calculate the ucb score for every child node in the list.
         choices_weights = [(c.q / c.n) + c_param * np.sqrt((2 * np.log(self.n) / c.n)) for c in self.children]
         self.best_action = self.parent_action[np.argmax(choices_weights)]
+        # Choose the child node which has the highest ucb score.
         return self.children[np.argmax(choices_weights)]
 
     def rollout_policy(self, possible_actions):
+        # randomly choose an action
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:105]: Overview:
            This method implements the rollout policy for a node during the Monte Carlo Tree Search.
            The rollout policy is used to determine the action taken during the simulation phase of the MCTS. In this case, the policy is to randomly choose an action from the list of possible actions.
        Arguments:
            - possible_actions: A list of all possible actions that can be taken from the current state.
        Return: 
            A randomly chosen action from the list of possible actions.

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: zoo/board_games/mcts_bot.py [hunk 5]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
...
 
     def rollout_policy(self, possible_actions):
+        # randomly choose an action 
         return possible_actions[np.random.randint(len(possible_actions))]
 
-
 class TwoPlayersMCTSNode(MCTSNode):
+    """
+    Overview:
+        The subclass, which inherits from the abstract base class, implements the specific methods required for a MCTS node.
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:111]: for a TwoPlayers *

File: zoo/board_games/mcts_bot.py [hunk 6]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
...
 
     def __init__(self, env, parent=None):
         super().__init__(env, parent)
         self._number_of_visits = 0.
+        # A default dictionary which sets the value to 0 for undefined keys.
         self._results = defaultdict(int)
         self._legal_actions = None
 
+    # Get all legal actions in current state from the environment object.
+    # 为什么不直接在initialize里定义啊？？？？？？？？？？？？？？？？？？？？？？？？？
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:122]: 可以改到__init__里面去哈

File: zoo/board_games/mcts_bot.py [hunk 7]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
...
     @property
     def legal_actions(self):
         if self._legal_actions is None:
             self._legal_actions = self.env.legal_actions
         return self._legal_actions
 
+
     @property
     def q(self):
         '''
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:132]: 统一注释格式"""

File: zoo/board_games/mcts_bot.py [hunk 8]
```diff
@@ -61,52 +77,70 @@ def backpropagate(self, reward):
...
+    # Get all legal actions in current state from the environment object.
+    # 为什么不直接在initialize里定义啊？？？？？？？？？？？？？？？？？？？？？？？？？
     @property
     def legal_actions(self):
         if self._legal_actions is None:
             self._legal_actions = self.env.legal_actions
         return self._legal_actions
 
+
     @property
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:130]: 整个方法优化为：    
```
@property
    def q(self):
        """
        This property represents the estimated value (Q-value) of the current node.

        self._results[1] represents the number of wins for player 1.
        self._results[-1] represents the number of wins for player 2.

        The Q-value is calculated depends on which player is the current player at the parent node,
        and is computed as the difference between the wins of the current player and the opponent.
        If the parent's current player is player 1, Q-value is the difference of player 1's wins and player 2's wins.
        If the parent's current player is player 2, Q-value is the difference of player 2's wins and player 1's wins.

        For example, if self._results[1] = 10 (player 1's wins) and self._results[-1] = 5 (player 2's wins):
        - If the parent's current player is player 1, then Q-value = 10 - 5 = 5.
        - If the parent's current player is player 2, then Q-value = 5 - 10 = -5.

        This way, a higher Q-value for a node indicates a higher win rate for the parent's current player.
        """

        # Determine the number of wins and losses based on the current player at the parent node.
        wins, loses = (self._results[1], self._results[-1]) if self.parent.env.current_player == 1 else (
        self._results[-1], self._results[1])

        # Calculate and return the Q-value as the difference between wins and losses.
        return wins - loses
```

File: zoo/board_games/mcts_bot.py [hunk 9]
```diff
@@ -125,17 +159,35 @@ def n(self):
...
+        # done is the bool flag representing whether the game is over
         return self.env.get_done_reward()[0]
 
     def rollout(self):
+        """
+        Overview:
+            Traverse the tree starting from the current node until reaching a terminal node.
+            Take action according to the rollout policy, such as random policy.
+        Returns:
+            reward (int):
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:186]: 按DI-engine的注释格式，参考[这里](https://github.com/opendilab/DI-engine/blob/main/ding/model/common/head.py)：- reward (:obj:`int`):

File: zoo/board_games/mcts_bot.py [hunk 10]
```diff
@@ -125,17 +159,35 @@ def n(self):
         return self._number_of_visits
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:159]: """
    This property represents the number of times the node has been visited during the search.
    """

File: zoo/board_games/mcts_bot.py [hunk 11]
```diff
@@ -125,17 +159,35 @@ def n(self):
...
 
     def expand(self):
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:161]:
```
   """
    Overview:
        This method expands the current node by creating a new child node.

        It pops an action from the list of legal actions, simulates the action to get the next game state,
        and creates a new child node with that state. The new child node is then added to the list of children nodes.
    """
```

File: zoo/board_games/mcts_bot.py [hunk 12]
```diff
@@ -125,17 +159,35 @@ def n(self):
...
         action = self.legal_actions.pop()
+        # simulate_action() returns a new environment object which reset the board and the current player flag.
         next_simulator_env = self.env.simulate_action(action)
+        # Create a new node object for the child node.
         child_node = TwoPlayersMCTSNode(next_simulator_env, parent=self)
         self.children.append(child_node)
         self.parent_action.append(action)
         return child_node
 
     def is_terminal_node(self):
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:172]:
```
    """
    Overview:
        This method checks if the current node is a terminal node.

        It uses the game environment's get_done_reward method to check if the game has ended.
    """
```

File: zoo/board_games/mcts_bot.py [hunk 13]
```diff
@@ -125,17 +159,35 @@ def n(self):
...
+        # reward = 0 when it is a tie
+        # reward = None when current node is not a teminal node
+        # done is the bool flag representing whether the game is over
         return self.env.get_done_reward()[0]
 
     def rollout(self):
+        """
+        Overview:
+            Traverse the tree starting from the current node until reaching a terminal node.
+            Take action according to the rollout policy, such as random policy.
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:184]: """
    This method performs a rollout (simulation) from the current node.

    It repeatedly selects an action based on the rollout policy and simulates the action until the game ends.
    The method then returns the reward of the game's final state.
    """

File: zoo/board_games/mcts_bot.py [hunk 14]
```diff
@@ -150,12 +202,20 @@ def rollout(self):
 
     def backpropagate(self, result):
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:203]: """
    This method performs backpropagation from the current node.

    It increments the number of times the node has been visited and the number of wins for the result.
    If the current node has a parent, the method recursively backpropagates the result to the parent.
    """

File: zoo/board_games/mcts_bot.py [hunk 15]
```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
...
                 # print('reward={}\n'.format(reward))
+                # backpropagate from the leaf node to the root node.
                 v.backpropagate(reward)
         # to select best child go for exploitation only
+        # 这里和算法理论里选择最高访问次数似乎不太一样？？？？？？？？？？？？？？？？？？
         return self.root.best_child(c_param=0.)
 
+    #
     def _tree_policy(self):
         """
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:269]: """
          Implement the tree policy for action selection during rollout.

          At each step, if the current node is not fully expanded, it expands.
          If it is fully expanded, it moves to the best child according to the tree policy.
          This continues until a terminal node is reached, which is then returned.

          Returns:
              The terminal node reached by following the tree policy.
          """

File: zoo/board_games/mcts_bot.py [hunk 16]
```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
             Returns the best children node, and can get action from Node.best_action_index.
         """
 
+        # The search cost is determined by either the maximum number of simulations or the longest simulation time.
         if simulations_number is None:
             assert (total_simulation_seconds is not None)
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:242]: # If no simulations number is provided, run simulations for the specified time

File: zoo/board_games/mcts_bot.py [hunk 17]
```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
...
             while True:
+                # get the leaf node.
                 v = self._tree_policy()
+                # rollout from the leaf node.
                 reward = v.rollout()
+                # backpropagate from the leaf node to the root node.
                 v.backpropagate(reward)
                 if time.time() > end_time:
                     break
         else:
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:253]: # If simulations number is provided, run the specified number of simulations

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: zoo/board_games/mcts_bot.py [hunk 18]
```diff
@@ -150,12 +202,20 @@ def rollout(self):
...
     def backpropagate(self, result):
         self._number_of_visits += 1.
+        # result is the index of the self._results list
+        # result = ±1 when player 1 wins/loses the game
         self._results[result] += 1.
         if self.parent:
             self.parent.backpropagate(result)
 
 
 class MCTSSearchNode(object):
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:212]: class名称改为MCTSSearch

File: zoo/board_games/mcts_bot.py [hunk 19]
```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
...
+                # choose a child node which has not been visited before
                 return current_node.expand()
             else:
                 current_node = current_node.best_child()
         return current_node
 
 
 class MCTSBot:
-
+    """
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:284]: """
         Initialize the Monte Carlo Tree Search Bot.

         Arguments:
             - ENV: The environment in which the MCTS bot operates.
             - cfg: Configuration parameters for the bot.
             - bot_name (str): The name of the bot.
             - num_simulation (int): Number of simulations to perform for each move. Default is 10000.
         """

File: zoo/board_games/mcts_bot.py [hunk 20]
```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
...
+        A robot which can use MCTS to make decision, choose an action to take.
+    """
+ 
     def __init__(self, ENV, cfg, bot_name, num_simulation=10000):
         self.name = bot_name
         self.num_simulation = num_simulation
         self.ENV = ENV
         self.cfg = cfg
 
     def get_actions(self, state, player_index):
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:295]: """
         Get the best action for a given game state.

         A new environment is created for simulations. The environment is reset to the given state.
         Then, MCTS is performed with the specified number of simulations to find the best action.

         Arguments:
             - state: The current game state.
             - player_index (int): The index of the current player.

         Returns:
             The best action determined by MCTS.
         """

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: zoo/board_games/mcts_bot.py [hunk 21]
```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
...
+        # 为什么把类作为属性，每次重新生成一个，而不是把环境实例直接作为属性？
+        # an instance of the ENV class.
         simulator_env = self.ENV(EasyDict(self.cfg))
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:298]: 确实应该在初始化中执行这一行哈，这里应该是之前写成了，你改好所有的后，在执行mcts_bot的测试文件哈

File: zoo/board_games/mcts_bot.py [hunk 22]
```diff
@@ -5,23 +5,39 @@
...
 
-
 class MCTSNode(ABC):
+    """
+    Overview:
+        An abstract base class.
+        Define basic methods for a Monte Carlo Tree Node.
+        The specific methods are implemented in the subclasses.
+        https://github.com/int8/monte-carlo-tree-search
+    """
```
[the remark this answers is not in the document]
Comment (HarryXuancy) [reply zoo/board_games/mcts_bot.py:15]: 已修改，并且在代码开头添加了整个代码的overview

File: zoo/board_games/mcts_bot.py [hunk 23]
```diff
@@ -5,23 +5,39 @@
...
+                        [1,1,0],   
+                        [2,0,0]]   
+                        where 0 stands for position that has not been played yet, 
+                        1 represents a position that has been played by player 1.
+                        2 represents a position taken by player 2.
+                    - env.players = [1,2], representing two players, player 1 and player 2.
+                    - env.start_player_index = 0 or 1. It will be reset every round.
+                    - env._current_player = env.players[env.start_player_index], representing which player 
+                        should make a move in this round.
+                - The methods in this object implement functionalities such as environment transitions and obtaining game results.
```
[the remark this answers is not in the document]
Comment (HarryXuancy) [reply zoo/board_games/mcts_bot.py:33]: 已修改

File: zoo/board_games/mcts_bot.py (hunk 1 above)
Comment (HarryXuancy) [reply zoo/board_games/mcts_bot.py:35]: 已修改

File: lzero/mcts/ptree/ptree_az.py [hunk 24]
```diff
@@ -18,12 +18,15 @@ class Node(object):
     Overview:
         The node base class for tree_search.
     """
-
+    # 把env作为节点的变量，或者单独分出去，这有没有什么说法
```
Comment (puyuan1996) [inline lzero/mcts/ptree/ptree_az.py:21]: lzero/mcts/ptree/ptree_az.py 按照原来的代码版本取消更改哈，ptree_az.py的注释优化新开一个PR

File: zoo/classic_control/cartpole/config/cartpole_muzero_config.py [hunk 25]
```diff
@@ -16,7 +16,7 @@
 # ==============================================================
 
 cartpole_muzero_config = dict(
-    exp_name=f'data_mz_ctree/cartpole_muzero_ns{num_simulations}_upc{update_per_collect}_rr{reanalyze_ratio}_seed0',
+    exp_name=f'data_mz_ctree/cartpole_muzero_ns{num_simulations}_upc{update_per_collect}_rr{reanalyze_ratio}_stopvalue150',
```
Comment (puyuan1996) [inline zoo/classic_control/cartpole/config/cartpole_muzero_config.py:19]: 这个文件取消更改哈，可以直接复制原仓库的代码

File: zoo/board_games/tictactoe/config/tictactoe_muzero_bot_mode_config.py [hunk 26]
```diff
@@ -87,4 +88,8 @@
 
 if __name__ == "__main__":
     from lzero.entry import train_muzero
-    train_muzero([main_config, create_config], seed=0, max_env_step=max_env_step)
```
Comment (puyuan1996) [inline zoo/board_games/tictactoe/config/tictactoe_muzero_bot_mode_config.py:90]: 这个文件取消更改哈，可以直接复制原仓库的代码

File: zoo/board_games/mcts_bot.py [hunk 27]
```diff
@@ -61,81 +78,162 @@ def backpropagate(self, reward):
...
-            wins = self._results[1]
-            loses = self._results[-1]
-
-        if self.parent.env.current_player == 2:
-            wins = self._results[-1]
-            loses = self._results[1]
-        # print("wins={}, loses={}".format(wins, loses))
+    def value(self):
+        """
+        Overview
```
Comment (puyuan1996) [inline zoo/board_games/mcts_bot.py:159]: Overview:

File: zoo/board_games/test_speed_win-rate_between_bots.py [hunk 28]
```diff
@@ -20,18 +27,34 @@
 
 
 def test_tictactoe_mcts_bot_vs_rule_bot_v0_bot(num_simulations=50):
+    """
+    Overview:
+        A tictactoe game between mcts_bot and rule_bot, where rule_bot take the first move.
+    Arguments:
+        -num_simulations (:obj:`int`): The number of the simulations required to find the best move.
```
Comment (puyuan1996) [inline zoo/board_games/test_speed_win-rate_between_bots.py:34]: 本文件里面的 `-num_simulations (:obj:`int`):` 替换为 `- num_simulations (:obj:`int`): `

**Pull request merged**